Issue 687 - Add Drawings Support

sebjf opened the pull request:
This fixes #687

#### Description
Adds support for processing drawings in a similar manner to 3d models, but where the drawing to process is provided via an existing revision node.

Specifically, this PR:

1. Makes `RevisionNode` a superclass and `ModelRevisionNode` & `DrawingRevisionNode` subclasses, with members appropriately scoped to each.
2. Adds the ability to get a path from an `fs` ref node, since that is how files will be provided for drawings as opposed to via the bouncer import config
3. Extends the `FileProcessor` superclass so that it can now be initialised to collect geometry data, SVG data, or both, from a single file (SVG data collected via the `DrawingImageInfo` type, which is analogous but not directly equivalent to `GeometryCollector`).
4. Adds the ability to import to SVGs, via `DrawingImageInfo`, to the DWG and DGN file processor subclasses. This uses the `SvgExportModule` which is included as part of ODA.
5. Introduces a new class `DrawingManager`, which is responsible for interacting with the database
6. Introduces a new class `DrawingImportManager`, which is responsible for selecting the right importer and using it to populate the `DrawingImageInfo`
7. Adds a new method, `processDrawingRevision`, to `RepoManipulator` & `RepoController`, that will acquire a file from a drawing revision, process it using the aforementioned importers, and update the revision node in the database.
8. Hooked up `processDrawingRevision` to a new command `processDrawing` in 3drepobouncerClient
9. Added unit tests for the new functionality

Additionally, changes to `bouncer_worker` include:

1. A new queue handler, `drawingQueueHandler.js`, has been added, as well as a new label to `queueLabel` for it.
2. Queue handler definitions have been decoupled from the config, with the associations now being made in `queueHandler.js`
3. `queueHandler.js` has been refactored slightly so that dereferencing (and validation) of labels to queue names from the config happens immediately in `connectToQueue` or `runNTasks`, and from there on in the queue names are used to resolve handlers, allowing queue initialisation and handling methods to be more generalized throughout.

carmenfan commented on tools/bouncer_worker/src/queues/modelQueueHandler.js:
```diff
@@ -61,18 +61,20 @@ Handler.onMessageReceived = async (cmd, rid, callback) => {
 
+		// Append process information with queue specific properties
+		const { size } = fs.statSync(file);
+		processInformation.FileType = file.split('.').pop().toString(), // filetype
```
could we do this for drawing stoo, unless i've mistakened I don't think we're adding this info in the drawing queue handler

@Charence we're goig to have similar model processing documents for 2D Drawings, they can be DWGs which means file type can't distinguish them (as we also have DWGs)

What do you propose? do we add an additional field or do something like "DWG (2D)" as the file type?

Charence replied:
@carmenfan @sebjf I think it depends if DWG is the only case where it can store both 2D and 3D and we want to distinguish if it came from a 2D or 3D file.

1. If `DWG (2D)` is kind of a one-off special case and we assume the person looking at the records knows if a format is 2D or 3D/doesn't care if it is 2D or 3D then I guess this could work.
2. If the person looking at the records wants to easily separate 2D vs 3D ones and we don't assume they know what format is 2D or 3D then I think we could store an additional field.

sebjf replied:
Hi @carmenfan , the reason its not done in the drawing queue is that bouncerWorker doesn't read the ref node, it only gets what's in the message in the queue (`database`,`model`,`user`,`revId`, for drawings).

Is the plan for the extension and size to be added to the rabbit mq message for drawings?

carmenfan replied:
ah right.

Ok let me just add it as part of the json since the backend has all that info

carmenfan replied:
@sebjf  will probably do it for both 3D and 2D if it makes a difference here

sebjf replied:
Hi @carmenfan @Charence,

There is an argument to `gatherProcessInformation` that takes a queue name as a label for the logs:

https://github.com/3drepo/3drepobouncer/blob/7f001fd8c1b125f318ee73c9df46348d2d44d60e/tools/bouncer_worker/src/queues/modelQueueHandler.js#L70

If this is distinct to each handler, then it is effectively the processing type, if it can be used that way @Charence? (I have set it to `DRAWINGQ`).

Please let me know where to find the updated schema for the queue messages when its done and I'll add the additional members to the drawings handler.

carmenfan replied:
> ah right.
> 
> Ok let me just add it as part of the json since the backend has all that info

@sebjf  this will come in as `format` and `size` as part of the json file

carmenfan commented on tools/bouncer_worker/src/lib/queueHandler.js:
```diff
@@ -54,22 +92,16 @@ const listenToQueue = (channel, queueName, prefetchCount, callback) => {
 		}
-		listenToQueue(channel, rabbitmq.model_queue, rabbitmq.model_prefetch, ModelQHandler.onMessageReceived);
-	}
+		listenToQueue(channel, queueName, rabbitmq.task_prefetch, handler.onMessageReceived);
```
Nice that it's in a loop instead of copy and pasting, but using `task_prefetch` everywhere will be problematic.

typically `task_prefetch` is set to 4 (we process up to 4 federation edits at any one time), but `model_prefetch` is set to 1 (only ever run 1 model import at any one time)

sebjf replied:
Hi @carmenfan , oops 🤦, I have updated it so the prefetch count comes from handler, so each `Handler` can get the correct entry from the config.

Do you want a new config entry for drawings, or just for it to use `model_prefetch` for now?

carmenfan replied:
yes if we can add one for drawing so it's all configurable that'll be good, thanks!

sebjf replied:
Done, the new member is called `drawing_prefetch`, e.g.

```
	"rabbitmq": {
		"host": "amqp://bouncerWorker:bouncerWorker@rabbitmq:5672",
		"worker_queue": "jobq",
		"callback_queue": "callbackq",
		"model_queue": "modelq",
		"drawing_queue": "drawingq",
		"task_prefetch": 4,
		"model_prefetch": 1,
		"drawing_prefetch": 2,
		"sharedDir": "D:/3drepo/local/efs/queue"
	},
```

sebjf commented on bouncer/src/repo/core/model/bson/repo_bson_ref.cpp:
```diff
@@ -34,10 +34,14 @@ std::string RepoRef::convertTypeAsString(const RefType &type) {
 }
 
-std::string RepoRef::getFileName() const {
+std::string RepoRef::getID() const {
```
Ironically the only place this is used is in the unit test... where it is [set to a string that is not a UUID](https://github.com/3drepo/3drepobouncer/blob/7f001fd8c1b125f318ee73c9df46348d2d44d60e/test/src/unit/repo/core/model/bson/ut_repo_bson_ref.cpp#L37)! 🙈

I have updated it to check for the type and added a unit test anyway though, in case someone does need it in the future and makes the same assumption..

sebjf commented on tools/bouncer_worker/src/lib/queueHandler.js:
```diff
@@ -168,41 +200,26 @@ const connectToRabbitMQ = async (autoReconnect, uponConnected) => {
-	} else {
-		logger.error(`Unrecognised queue type: ${specificQueue}. Expected [job|model]`, logLabel);
-		exitApplication();
+	let queueNames = [];
```
Turns out it does complain if its run properly! Fixed.

sebjf commented on tools/bouncer_worker/src/lib/queueHandler.js:
```diff
@@ -168,41 +200,26 @@ const connectToRabbitMQ = async (autoReconnect, uponConnected) => {
+		queueNames.push(getQueueName(specificQueue));
 	}
+	else{
+		for(const label in queueLabel){
```
Done!

carmenfan commented:
@sebjf  another one:

The backend is storing the `rFile` reference as UUID so when I run the whole pipeline it's struggling to locate the file.

![image](https://github.com/user-attachments/assets/e8e84859-b2e7-4b18-aa39-4970e48e3938)


Ideally we'd like to keep it as LUUID, but let me know if this is going to be super problematic. If you want to try it ISSUE_5090 on3drepo.io is kind of hooked up (i don't know for 100% as I couldn't trigger a successful drawing process yet!) You will need to add drawing_queue config on 3drepo.io for it to work (similar to what you did on bouncer worker)

sebjf commented:
Hi @carmenfan,

> can you also update the config_example

Yes it would be helpful to update the one in github and not just my local! 🙈 

> yargs helper also need updating I think

I have done another pass and updated all the help strings too.

> The backend is storing the rFile reference as UUID

`DrawingRevisionNode` does assume the `rFile` entries are UUIDs, so not sure why its not working... 

(This for example is what I have been running bouncer against...)

![image](https://github.com/user-attachments/assets/5b8c8c46-4582-4d98-be79-9c08464b245e)

I will check out 5090 and see if I can get it to run end-to-end!

carmenfan commented:
> DrawingRevisionNode does assume the rFile entries are UUIDs, so not sure why its not working.

@sebjf  ah in that case it may be my fault... do let me know if the issue is on the backend 😆

sebjf commented:
> do let me know if the issue is on the backend 😆

OK looked into it and its not the backend! The `DrawingRevisionNode` assumes the `rFile` is a UUID... `getFileRef` however does *not*!

I will make an overload for callers who know they will be looking for UUIDs in FileManager and updating the drawings processing code to use it..

sebjf commented:
Hi @carmenfan, i've added a `getFileRef` overload that can now pick up on the document from the backend. However I also note that `FileManager::uploadFileAndCommit` will create ref nodes for the svgs with `_id`'s as strings, not UUIDs.

Would you like me to create an overload for this & have drawings use UUIDs here too?

carmenfan commented:
>  i've added a getFileRef overload that can now pick up on the document from the backend. 
@sebjf 
Thank you 😄 

> Would you like me to create an overload for this & have drawings use UUIDs here too?
It would be ideal if we could (just for a better/smaller index on the database), but thanks to javascript, backend understands it either way so it's not too big of a trouble.

sebjf commented:
> It would be ideal if we could

Yes I think its straightforward to make an overload of `uploadFileAndCommit`. 

I did look just now at possibly making it so all ref nodes were UUIDs by default, if the backend would be fine with some projects mixing the key types.

I think it should be possible as the only thing iirc that used non-UUID names was the previous `MeshNode` schema which would append strings like "_vertices"; but the new `_blobRef` schema no longer does that.

JavaScript would be fine, and for the few cases bouncer reads files (i.e. when loading binary mapped RepoBSON nodes), we could have `getFile` try one type first and fallback to the other, to emulate JavaScript.

I think if creating the overload is simple enough though, we should just do that for now, then make a maintenance ticket to change the other calling locations over to use it in the future, as this is getting out of scope now!

carmenfan commented:
@sebjf up to you, I think ideally we want to convert all strings to UUIDs, but timing wise I'll leave it to your judgement

As a side note I would also like to make it so the 3D models read the file from revision node like drawings and stop getting bouncer to create a rev node (so we can start tracking individual progress instead of piling it into the model settings! - there is a ticket for this but I can't find it right now) but this is def 100% out of scope 😆

sebjf commented:
> As a side note I would also like to make it so the 3D models read the file from revision node...there is a ticket for this but I can't find it right now

Is it this one? https://github.com/3drepo/3D-Repo-Product-Team/issues/446

(I want to add a note to whichever it is to deprecate the ref-node-id-as-string as part of it! 😆)

carmenfan commented:
> @carmenfan
> 
> > As a side note I would also like to make it so the 3D models read the file from revision node...there is a ticket for this but I can't find it right now
> 
> Is it this one? [3drepo/3D-Repo-Product-Team#446](https://github.com/3drepo/3D-Repo-Product-Team/issues/446)
> 
> (I want to add a note to whichever it is to deprecate the ref-node-id-as-string as part of it! 😆)

Yes! Good job for finding it, I tried yesterday and failed miserably 🙈

sebjf commented:
Hi @carmenfan, did you say that size and format had been added to the queue message?

It doesn't appear to be so in 5090, though I am probably looking in the wrong place!

![image](https://github.com/user-attachments/assets/3d87f945-be70-4f65-9365-14d8bc131a97)

carmenfan commented:
@sebjf 

It should be added as part of the `queueMeta`, which is the `data` object in the function you're looking at
![image](https://github.com/user-attachments/assets/2537eddd-8eec-40ba-9ed8-9510e558c6ec)


And eventually gets writting inside importParams.json  (equivalent to `<corId>.json` file in the 3D Model implementation)
![image](https://github.com/user-attachments/assets/efb852c2-e0a5-4ea0-a3e0-37320e61c95b)

sebjf commented:
Thank you!

sebjf commented:
@carmenfan,

This is now receiving a post request like so,

`https://3drepo.lan/api/v5/teamspaces/sebjf/projects/59284d10-4b00-11ed-ac46-93641d60e710/drawings/17e4db45-bf25-4b87-a5ad-00a2e781217f/revisions?key=04befbd4f356b2117b6a7ddc344e6f97`

And processing it successfully. 

We might want to wait for the endpoint to get the SVGs to make sure it can be read as well, but if that works all the changes should be addressed now!

carmenfan reviewed (changes requested):
@sebjf  There's a type descrepancy on the id of the image:

tjhe entry in .history is written as a UUID:
![image](https://github.com/user-attachments/assets/1c2d64b0-e48b-4a25-bc2e-7ca581a1bcf4)

But the ID in `.ref` is in string
![image](https://github.com/user-attachments/assets/f68ab55b-04da-4742-acd8-6b0161d64f43)

JS will just pass it as is to the database query so mongo will return with no entry. So we either need both to be string or both to be UUID 😢 


Also, can we conver the extension on the name to be .svg please?

carmenfan commented:
@sebjf another thing, just checking out the svg against the same drawing exported to PDF:
![image](https://github.com/user-attachments/assets/14ca2d75-7dcb-419b-b94e-fb44b4ccb072)


The annotations are coming out in weird colours, sometimes it's it's obscured as well because the background colour is the same as the font colour:
![image](https://github.com/user-attachments/assets/e743b193-af3e-41cf-9387-150ef8573855)

carmenfan commented:
> @sebjf another thing, just checking out the svg against the same drawing exported to PDF, this is what the PDF looks like: ![image](https://private-user-images.githubusercontent.com/11945337/356897976-14ca2d75-7dcb-419b-b94e-fb44b4ccb072.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM0MDA4OTAsIm5iZiI6MTcyMzQwMDU5MCwicGF0aCI6Ii8xMTk0NTMzNy8zNTY4OTc5NzYtMTRjYTJkNzUtN2RjYi00MTliLWI5NGUtZmI0NGI0Y2NiMDcyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODExVDE4MjMxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI1OGE1YzYzYWMyMmYyNTkxODFlOGVhY2ZmZmZhZTU1MGQzZmQ4Yzg3ZGJmZGZmYzQxNGM2N2RjNTBkYTNiOWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.RuDmB9JXw5GbiewGXFb0MVu4Rbv-To4BGIXmKNsxcZw)
> 
> And this is the DWG. The annotations are coming out in weird colours, sometimes it's it's obscured as well because the background colour is the same as the font colour: ![image](https://private-user-images.githubusercontent.com/11945337/356898060-e743b193-af3e-41cf-9387-150ef8573855.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM0MDA4OTAsIm5iZiI6MTcyMzQwMDU5MCwicGF0aCI6Ii8xMTk0NTMzNy8zNTY4OTgwNjAtZTc0M2IxOTMtYWYzZS00MWNmLTkzODctMTUwZWY4NTczODU1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODExVDE4MjMxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVhYjYwOTIxZDhiOGMxM2M2NThiYjVjNDEyNjQ2NGMxOTk3YzVhMWZhZDg1ZDM3NzAzN2I5Njk2YWY5YmZjNzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Y5DQdFrkK5cGaooNC1mcu4ujn3A9ZB-g8tBgBxWOEIw)

Opening it the ODA app i can see it's assuming a black background... I wonder if this is a setting we can tweak...
![image](https://github.com/user-attachments/assets/0d8267b4-7738-4a2c-a203-bc8100ee677b)

sebjf commented:
Hi @carmenfan,

> @sebjf There's a type descrepancy on the id of the image:

Gah, missed it after the refactor - fixed!

> Also, can we conver the extension on the name to be .svg please?

Done!

![image](https://github.com/user-attachments/assets/346a047c-91a2-40d4-bf37-11fb8a6f41eb)

I noticed though that `model` in the revision node created by .io is a string too. Is this intentional? (bouncer reads it and assumes its a UUID so I'll need to update it if so)

sebjf commented:
Hi @carmenfan,

> Opening it the ODA app i can see it's assuming a black background... I wonder if this is a setting we can tweak...

Yes, the `OdGiContextForDbDatabase` has a few settings. I have updated the palette background and that seems to have worked. I also updated the minimum line with and it looks a little closer to Revit too.

carmenfan commented:
> Hi @carmenfan,
> 
> > @sebjf There's a type descrepancy on the id of the image:
> 
> Gah, missed it after the refactor - fixed!
> 
> > Also, can we conver the extension on the name to be .svg please?
> 
> Done!
> 
> ![image](https://private-user-images.githubusercontent.com/5058369/357336881-346a047c-91a2-40d4-bf37-11fb8a6f41eb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM1NDAxODEsIm5iZiI6MTcyMzUzOTg4MSwicGF0aCI6Ii81MDU4MzY5LzM1NzMzNjg4MS0zNDZhMDQ3Yy05MWEyLTQwZDQtYmYzNy0xMWZiOGE2ZjQxZWIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MTNUMDkwNDQxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWU3NTMxYzAyZWE3MWI5NjAzYTczNjg1NmI0M2QxMzM2MzQxZDE5YWI5ZDJhY2VlYTc4YWRmZDI1ODM5NzliOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.-RqRBubNxaJanY1Za4nZ2wgBKajO1NTaE2ydTxTSRdE)
> 
> I noticed though that `model` in the revision node created by .io is a string too. Is this intentional? (bouncer reads it and assumes its a UUID so I'll need to update it if so)

model is stored as a string because of legacy reason (if you open the settings collection in a db you'll see they're all string ID, and it'll be difficult to store it as LUUID in one place and string in other as they're usually passed around and queried everywhere)

I do want to convert to UUID in the future but it's going to take a big migration script 🤦... so future work 😆

sebjf commented:
> I do want to convert to UUID in the future but it's going to take a big migration script 🤦... so future work 😆

Got it - bouncer has been updated to consider it a string!

sebjf commented:
(@carmenfan , something I just remembered was that ODA fails to set the `width` and `height` tags of its SVGs, which Chromium at least needs to render anything sensible, so we should leave this open until that's added at least.)

carmenfan commented:
> (@carmenfan , something I just remembered was that ODA fails to set the `width` and `height` tags of its SVGs, which Chromium at least needs to render anything sensible, so we should leave this open until that's added at least - we'll probably need to add those ourselves to the string before its saved.)

@sebjf  not 100% sure what this means 😆  . Are we waiting on a fix from ODA? or do we need add this ourselves as part of this branch?

carmenfan reviewed (dismissed):
LGTM - just need to double check with you what the SVG height/width entails 😆

sebjf commented:
> @sebjf not 100% sure what this means 😆 . Are we waiting on a fix from ODA? or do we need add this ourselves as part of this branch?

Sorry end of the day post 😆 

SVG files have both a `viewBox`, and `width` and `height` tags. I am not sure why they need all three since `viewBox` should do the job. But for whatever reason, they do. ODA however doesn't write the `width` and `height` properties. 

This should not be problem in theory because the `viewBox` is known, and the true with and height of the renderer is set by the DOM. However, in practice odd things happen if they aren't present. For example, it seems Firefox assumes the fields are present and there are many bugs like [this](https://github.com/openlayers/openlayers/issues/14237) from downstream libraries that use SVGs with it.

At first I was just going to make sure our 2D viewer could handle SVGs without the tags, but given Firefox's issues, I think the best solution is just to add the fields on export. We know the width and height because we set up the view properties ourselves.

This is something we'd do in bouncer; we wouldn't try and get ODA to modify their exporter.

I will do it today as part of this ticket, in case we want to merge before Milestone 5.

sebjf commented:
Hi @carmenfan,

Added the attributes. There are a couple more magic numbers/strings than I usually like but not sure its worth creating an svg helper header yet - I've updated the unit tests in case the svg exporter changes with the ODA upgrade. 
(I guess we could put a function in the FileProcessor base class, let me know if thats preferred!)

carmenfan commented:
@sebjf thanks!

Probably prefer it to be in the base class but not really a big issue 😆
